### PR TITLE
Improved ePub cover import

### PR DESCRIPTION
**Pull request by aristippe**
Should now handle everything in spec and some common cases.
- in guide as type "cover" or non-standard title = "Cover"
- metadata cover content points to cover file itself rather than manifest item
- if there's no cover, check for title page in guide as type 'title-page" or "tp"
- ePub 3 manifest item "cover-image"

Added a new ePub sample for the cover being specified in the guide, which seems the only thing ePub 2 requires. Should take care of #26 except for the rare non-spec covers if we someday want to deal with that. It's worth testing on a folder of many ePubs if you can, and then looking through the OPF of ones that lack covers.

### CODE REVIEW

File: books/epub.py [hunk 1]
```diff
@@ -149,13 +150,58 @@ def get_info(self):
         return self._info
 
     def get_cover_image_path(self):
```
Comment (sinergatis) [inline books/epub.py:153]: Can you please add some try..catch blocks when returning cover_image (and not `None`), that basically check that the file actually exists `os.path.exists`, and if not, return `None`? This would ward us against epubs that seem to point to a cover, but don't really have the image file, and basically save some trouble on functions that call this one.

File: books/epub.py [hunk 2]
```diff
@@ -149,13 +150,58 @@ def get_info(self):
...
+        """
+        if not self._info.cover_image:
             return None
 
+        self._unzip_file(self._basepath + self._info.cover_image)
+        cover_image = os.path.join(self._tempdir,
+                                   self._basepath,
+                                   self._info.cover_image)
+
+        if self._info.cover_image.endswith(('.gif', '.jpg', '.jpeg', '.png')):
```
Comment (sinergatis) [inline books/epub.py]: Very-minor-thing: can you make the comparison case insensitve, using `self._info.cover_image.lower` or similar?

File: books/epubinfo.py [hunk 3]
```diff
@@ -134,18 +134,59 @@ def _get_subject(self):
...
                     if node.attrib['id'] == meta_content_cover:
                         cover = node.attrib['href']
+            # In case meta tag refers to cover itself
+            if cover is None:
+                cover = meta_content_cover
+
+        # Guide
+        if cover is None:
+            for node in self._tree.iter():
+                if node.tag == '{http://www.idpf.org/2007/opf}reference':
```
Comment (sinergatis) [inline books/epubinfo.py:141]: Since all the next comparisons follow the same structure, can you merge them into one to avoid traversing the tree a bunch of times, ie:

``` python
if cover is None:
    for node in self._tree.iter():
        # epub foo
        if node.tag = ...... :
            if node.attrib['type'] == 'cover':
                cover = ....
                break
        # epub bar
        elif node.tag = ....
```

The break is basically for exiting the `for` when a match is found (not sure if any format/spec allows several elements of the same cover type, but just in case).

File: books/tests/sample_epubs.py [hunk 4]
```diff
@@ -15,10 +15,14 @@
...
                         os.path.join(RSRC_DIR, 'epub30-spec.epub'),
-                        True, False),
+                        True, True),
              SampleEpub('figure-gallery', 'figure-gallery-bindings.epub',
                         os.path.join(RSRC_DIR, 'figure-gallery-bindings.epub'),
                         True, True),
+             # Cover type = guide
+             SampleEpub('israelsailing', 'israelsailing.epub',
+                        os.path.join(RSRC_DIR, 'israelsailing.epub'),
+                        True, True),
```
Comment (sinergatis) [inline books/tests/sample_epubs.py:29]: Nice, a new testing epub is always nice! Can you update the `README.md` with basic info about the new `israelsailing` entry?

Also, if `epub30-spec` indeed has a cover (I blindy assumed it did not, as the previous version of the parser could not find it), would you be so king of trying to find a sample epub _without_ cover, and add it to the repo? This would help keeping the tests explicitly test the files without covers.

**Comment by sinergatis** [comment]
It's looking very nice, except for a couple of minor nitpicks aimed towards not having to revise it again later! I don't really have a significant amount of epubs that I know their cover/no cover status or their internals, so basically I'll trust your testing efforts in that regard - so feel free to merge after revising the minor comments if the tests pass!

**Comment by aristippe** [comment]
Thanks for the suggestions. Please take a look whenever you can. I re-organized the order of checking cover types a bit, added another, and refactored the variable names a bit. I believe according to PEP8, it's required to handle the exceptions in someway; when unzipping a non-existent file, it's logged and will show up during testing; then None is returned so the ePub will import successfully without a cover.

There is indeed a lot of repetitive code in Epubinfo and possibly a nice way of dealing with that is abstracting lxml into helper functions. See how Calibre does it[1] for one example. This was my first time playing around with lxml and it took a bit of time to figure out; it works well enough for now so I'm ok with it as is, though for anyone that wants nice code, it's worth looking again at someday.

[1] https://github.com/kovidgoyal/calibre/blob/master/src/calibre/ebooks/metadata/opf2.py

File: .gitignore [hunk 5]
```diff
@@ -1,5 +1,6 @@
 database.db
 *pyc
+books/migrations/*
```
Comment (sinergatis) [inline .gitignore:3]: A note: migration files _are_ supposed to be part of the distribution (even if they look rather ugly) and needed for the complete package, in order to allow users to update the DB schema and data seamlessly when a new version modifies the models (something [like this](http://stackoverflow.com/questions/28035119/should-i-be-adding-the-django-migration-files-in-the-gitignore-file)). I hope the line was meant to prevent accidental inclusion or problems when several developers are working on features that modify models - leaving it there for that reason is fine, and maybe save some trouble indeed.

**Comment by sinergatis** [comment]
Excellent, thanks!

I also do feel `epubinfo` and `epub` are in need of revision up to some degree indeed, for readability, robustness, and niceness in general. If you feel like improving them that's great, but it's probably a good idea to put some effort in looking at existing ePub parsing packages as using them would probably be easier and give us more ePub support "for free" and save some pain along the way.

So, in short ... merged! :dancers:

File: .gitignore (hunk 5 above)
Comment (aristippe) [inline .gitignore:3]: I thought perhaps migrate would take care of it, from whatever version the person is upgrading from, but now that I think about it, it makes sense for makemigrations which is probably why I have those files. I was experimenting with making authors many-to-many (so it can include multiple authors, contributors, per spec). I only tried it for a few moments and then noticed likely other things needed changing as well.

Comment (sinergatis) [inline .gitignore:3]: No worries, just making sure it was included for the right reasons!

I'd actually love to discuss some modifications to the models (in order to make them all at once, or at least the ones we see more essential), as it makes much sense to get them sorted out as soon as possible in order not to need revising future features once again when that's done. I'll open an issue!

Comment (aristippe) [inline .gitignore:3]: Awesome@ agree models are something to hopefully take a look at sometime soon. :)
